src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java

```diff
@@ -76,38 +89,90 @@ protected void start(Map<String, String> props, ElasticsearchClient client) {
     }
 
-    this.client = client != null ? client : new ElasticsearchClient(config, reporter);
+    this.client = client != null ? client
+            : new ElasticsearchClient(config, reporter, offsetTracker);
 
     log.info("Started ElasticsearchSinkTask. Connecting to ES server version: {}",
         getServerVersion());
   }
 
+  private static class PartitionPauser {
+
+    private final SinkTaskContext context;
+    private final BooleanSupplier pauseCondition;
+    private final BooleanSupplier resumeCondition;
+    private boolean partitionsPaused;
+
+    public PartitionPauser(SinkTaskContext context,
+                           BooleanSupplier pauseCondition,
+                           BooleanSupplier resumeCondition) {
+      this.context = context;
+      this.pauseCondition = pauseCondition;
+      this.resumeCondition = resumeCondition;
+    }
+
+    /**
+     * Resume partitions if they are paused and resume condition is met.
+     * Has to be run in the connector thread.
+     */
+    private void tryResumePartitions() {
+      if (partitionsPaused) {
+        if (resumeCondition.getAsBoolean()) {
+          log.debug("Resuming all partitions");
+          context.resume(context.assignment().toArray(new TopicPartition[0]));
+          partitionsPaused = false;
+        } else {
+          context.timeout(100);
+        }
+      }
+    }
+
+    /**
+     * Pause partitions if they are paused and resume condition is met.
+     * Has to be run in the connector thread.
+     */
+    private void tryPausePartitions() {
+      if (!partitionsPaused && pauseCondition.getAsBoolean()) {
+        log.debug("Pausing all partitions");
+        context.pause(context.assignment().toArray(new TopicPartition[0]));
+        context.timeout(100);
+        partitionsPaused = true;
+      }
+    }
+  }
+
   @Override
   public void put(Collection<SinkRecord> records) throws ConnectException {
     log.debug("Putting {} records to Elasticsearch.", records.size());
+
+    client.throwIfFailed();
+    partitionPauser.tryResumePartitions();
+
     for (SinkRecord record : records) {
+      OffsetState offsetState = offsetTracker.addPendingRecord(record);
+
       if (shouldSkipRecord(record)) {
         logTrace("Ignoring {} with null value.", record);
         reportBadRecord(record, new ConnectException("Cannot write null valued record."));
         continue;
       }
 
       logTrace("Writing {} to Elasticsearch.", record);
-
-      ensureIndexExists(convertTopicToIndexName(record.topic()));
-      checkMapping(record);
-      tryWriteRecord(record);
+      tryWriteRecord(record, offsetState);
     }
+    partitionPauser.tryPausePartitions();
   }
 
   @Override
-  public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
-    log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
+  public Map<TopicPartition, OffsetAndMetadata> preCommit(Map<TopicPartition,
+          OffsetAndMetadata> currentOffsets) {
```

Comment:
Nit: the lines splits here are unconventional. Should be:
```suggestion
  public Map<TopicPartition, OffsetAndMetadata> preCommit(
          Map<TopicPartition, OffsetAndMetadata> currentOffsets
  ) {
```